src/cargo/core/resolver/resolve.rs

```diff
@@ -23,15 +23,33 @@ pub struct Resolve {
     /// from `Cargo.toml`. We need a `Vec` here because the same package
     /// might be present in both `[dependencies]` and `[build-dependencies]`.
     graph: Graph<PackageId, Vec<Dependency>>,
+    /// Replacements from the `[replace]` table.
     replacements: HashMap<PackageId, PackageId>,
+    /// Inverted version of `replacements`.
     reverse_replacements: HashMap<PackageId, PackageId>,
+    /// An empty `HashSet` to avoid creating a new `HashSet` for every package
+    /// that does not have any features, and to avoid using `Option` to
+    /// simplify the API.
     empty_features: HashSet<String>,
+    /// Features enabled for a given package.
     features: HashMap<PackageId, HashSet<String>>,
+    /// Checksum for each package. A SHA256 hash of the `.crate` file used to
+    /// validate the correct crate file is used. This is `None` for sources
+    /// that do not use `.crate` files, like path or git dependencies.
     checksums: HashMap<PackageId, Option<String>>,
+    /// "Unknown" metadata. This is a collection of extra, unrecognized data
+    /// found in the `[metadata]` section of `Cargo.lock`, preserved for
+    /// forwards compatibility.
     metadata: Metadata,
+    /// `[patch]` entries that did not match anything, preserved in
+    /// `Cargo.lock` as the `[[patch.unused]]` table array.
+    /// TODO: *Why* is this kept in `Cargo.lock`? Removing it doesn't seem to
```

Comment:
I couldn't figure out the purpose of `unused_patches`, any ideas?

Reply:
Looks like it was added in #4123 if that helps...

Reply:
One of the features of `[patch]` is that it ends up needing to resolve all the dependencies all the time. This means that if we didn't record unused patch entries in `Cargo.lock` then every time you type `cargo build` Cargo would "Updating registry ..." and take a lot longer on each build. (it's not guaranteed this would happen but it's very likely this would happen)

By recording unused `[patch]` entries in the lock file we can ensure that all subsequent `cargo build` have a guaranteed resolution of what's in the manifest (so long as it hasn't changed) so we won't ever have to re-update the registry and we can get nice, quick, deterministic builds.

You can sort of think of it like adding an entry to `[dependencies]` but you don't actually use it, just in this case we know whether it's used or not.

Reply:
Ah, that makes sense! Thanks for the explanation.